ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,749 @@
+	pub proposals: RwLock<HashMap<Address, VoteType>>,
+	pub signer: RwLock<Option<Box<EngineSigner>>>,
+	pub step_service: Option<Arc<StepService>>,
+}
+
+impl Clique {
+	/// Initialize Clique engine from empty state.
+	pub fn new(our_params: CliqueParams, machine: EthereumMachine) -> Result<Arc<Self>, Error> {
+		let mut engine = Clique {
+			epoch_length: our_params.epoch,
+			period: our_params.period,
+			client: Default::default(),
+			block_state_by_hash: RwLock::new(LruCache::new(STATE_CACHE_NUM)),
+			proposals: Default::default(),
+			signer: Default::default(),
+			machine,
+			step_service: None,
+		};
+
+		let res = Arc::new(engine);
+
+		if our_params.period > 0 {
+			engine.step_service = Some(StepService::start(Arc::downgrade(&res) as Weak<Engine<_>>));
+		}
+
+		Ok(res)
+	}
+
+	#[cfg(test)]
+	/// Initialize test variant of `CliqueEngine`,
+	/// Note we need to `mock` the miner and it is introduced to test block verification to trigger new blocks
+	/// to mainly test consensus edge cases
+	pub fn with_test(epoch_length: u64, period: u64) -> Self {
+		use spec::Spec;
+
+		Self {
+			epoch_length,
+			period,
+			client: Default::default(),
+			block_state_by_hash: RwLock::new(LruCache::new(STATE_CACHE_NUM)),
+			proposals: Default::default(),
+			signer: Default::default(),
+			machine: Spec::new_test_machine(),
+			step_service: None,
+		}
+	}
+
+	fn sign_header(&self, header: &Header) -> Result<(Signature, H256), Error> {
+
+		match self.signer.read().as_ref() {
+			None => {
+				Err(EngineError::RequiresSigner)?
+			}
+			Some(signer) => {
+				let digest = header.hash();
+				match signer.sign(digest) {
+					Ok(sig) => Ok((sig, digest)),
+					Err(e) => Err(EngineError::Custom(e.into()))?,
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		debug_assert_eq!(header.number() % self.epoch_length, 0);
+
+		let mut state = CliqueBlockState::new(
+			extract_signers(header)?);
+
+		state.calc_next_timestamp(header, self.period);
+
+		Ok(state)
+	}
+
+	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		self.block_state_by_hash.write().get_mut(hash).cloned()
+	}
+
+	/// Get `CliqueBlockState` for given header, backfill from last checkpoint if needed.
+	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		let mut block_state_by_hash = self.block_state_by_hash.write();
+		if let Some(state) = block_state_by_hash.get_mut(&header.hash()) {
+			return Ok(state.clone());
+		}
+		// If we are looking for an checkpoint block state, we can directly reconstruct it.
+		if header.number() % self.epoch_length == 0 {
+			let state = self.new_checkpoint_state(header)?;
+			block_state_by_hash.insert(header.hash(), state.clone());
+			return Ok(state);
+		}
+		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
```

Comment:
This path may be expensive. I would suggest that we store some information in the block metadata on disk. (Metadata support was removed in a previous commit, but it should be relatively easy to cherry-pick it back).

Reply:
related: https://github.com/goerli/parity-goerli/issues/63

I'll take a look into this.